v3/sqlite/sqlite.go

```diff
@@ -0,0 +1,256 @@
+	if results.IsErr() {
+		return nil, toError(results.Err())
+	}
+	d.conn = &conn{
+		spinConn: *results.OK(),
+	}
+	return d.conn, nil
+}
+
+// Close closes the connection to the database.
+func (d *connector) Close() error {
+	if d.conn != nil {
+		d.conn.Close()
+	}
+	return nil
+}
+
+type rows struct {
+	columns []string
+	pos     int
+	len     int
+	rows    [][]any
+}
+
+var _ driver.Rows = (*rows)(nil)
+
+// Columns return column names.
+func (r *rows) Columns() []string {
+	return r.columns
+}
+
+// Close closes the rows iterator.
+func (r *rows) Close() error {
+	r.rows = nil
+	r.pos = 0
+	r.len = 0
+	return nil
+}
+
+// Next moves the cursor to the next row.
+func (r *rows) Next(dest []driver.Value) error {
+	if !r.HasNextResultSet() {
+		return io.EOF
+	}
+	for i := 0; i != len(r.columns); i++ {
+		dest[i] = driver.Value(r.rows[r.pos][i])
+	}
+	r.pos++
+	return nil
+}
+
+// HasNextResultSet is called at the end of the current result set and
+// reports whether there is another result set after the current one.
+func (r *rows) HasNextResultSet() bool {
+	return r.pos < r.len
+}
+
+// NextResultSet advances the driver to the next result set even
+// if there are remaining rows in the current result set.
+//
+// NextResultSet should return io.EOF when there are no more result sets.
+func (r *rows) NextResultSet() error {
+	if r.HasNextResultSet() {
+		r.pos++
+		return nil
+	}
+	return io.EOF // Per interface spec.
+}
+
+type stmt struct {
+	conn  *conn
+	query string
+}
+
+var _ driver.Stmt = (*stmt)(nil)
+var _ driver.ColumnConverter = (*stmt)(nil)
+
+// Close closes the statement.
+func (s *stmt) Close() error {
+	return nil
+}
+
+// NumInput returns the number of placeholder parameters.
+func (s *stmt) NumInput() int {
+	// Golang sql won't sanity check argument counts before Query.
+	return -1
+}
+
+func toRow(row []sqlite.Value) []any {
+	ret := make([]any, len(row))
+	for i, v := range row {
+		switch v.String() {
+		case "integer":
+			ret[i] = *v.Integer()
+		case "real":
+			ret[i] = *v.Real()
+		case "text":
+			ret[i] = *v.Text()
+		case "blob":
+			// TODO: check this
+			ret[i] = *v.Blob()
+		case "null":
+			ret[i] = nil
+		default:
+			panic("unknown value type")
+		}
+	}
+	return ret
+}
+
+func toWasiValue(x any) sqlite.Value {
+	switch v := x.(type) {
+	case int:
+		return sqlite.ValueInteger(int64(v))
+	case int64:
+		return sqlite.ValueInteger(v)
+	case float64:
+		return sqlite.ValueReal(v)
+	case string:
+		return sqlite.ValueText(v)
+	case []byte:
+		return sqlite.ValueBlob(cm.ToList([]uint8(v)))
+	case nil:
+		return sqlite.ValueNull()
+	default:
+		panic("unknown value type")
+	}
+}
+
+// Query executes a query that may return rows, such as a SELECT.
+func (s *stmt) Query(args []driver.Value) (driver.Rows, error) {
+	params := make([]sqlite.Value, len(args))
+	for i := range args {
+		params[i] = toWasiValue(args[i])
+	}
+	results := s.conn.spinConn.Execute(s.query, cm.ToList(params))
```

Comment:
There is now a shortcut for checking if result is an error. something like:

```
	okresult, err, isError := s.conn.spinConn.Execute(s.query, cm.ToList(params)).Result()
	if isError {
		return nil, toError(&err)
	}
```

Reply:
Nice! Updated